src/core/cluster.py

```diff
@@ -0,0 +1,211 @@
+import logging
+from typing import TYPE_CHECKING, Literal, Set
+
+from ops.model import Relation
+from src.core.models import ZKClient, ZKCluster, ZKServer
+
+if TYPE_CHECKING:
+    from charm import ZooKeeperCharm
+
+
+logger = logging.getLogger(__name__)
+
+# TODO: move to some better literals place
+SUBSTRATE = Literal["vm", "k8s"]
+PEER = "cluster"
+REL_NAME = "zookeeper"
+
+
+class ClusterState:
+    def __init__(self, charm: "ZooKeeperCharm", substrate: SUBSTRATE):
+        self.charm = charm
+        self.substrate: SUBSTRATE = substrate
+
+    # --- RAW RELATION ---
+
+    @property
+    def peer_relation(self) -> Relation | None:
+        return self.charm.model.get_relation(PEER)
+
+    @property
+    def client_relations(self) -> Set[Relation]:
+        return set(self.charm.model.relations[REL_NAME])
+
+    # --- CORE COMPONENTS---
+
+    @property
+    def unit_server(self) -> ZKServer:
+        return ZKServer(
+            relation=self.peer_relation, component=self.charm.unit, substrate=self.substrate
+        )
```

Comment:
`self.charm` is only used to access `unit` and `app` here. I think you can inherit from `ops.framework.Object` and use `self.model` instead? This way you can remove the `ZooKeeperCharm` typing
```suggestion
from ops.framework import Framework, Object

class ClusterState(Object):
    def __init__(self, charm: Framework | Object, substrate: SUBSTRATE):
        super().__init__(parent=charm, key="charm_state")
        self.substrate: SUBSTRATE = substrate
   
    # --- RAW RELATION ---

    @property
    def peer_relation(self) -> Relation | None:
        return self.model.get_relation(PEER)

    @property
    def client_relations(self) -> Set[Relation]:
        return set(self.model.relations[REL_NAME])

    # --- CORE COMPONENTS---

    @property
    def unit_server(self) -> ZKServer:
        return ZKServer(
            relation=self.peer_relation, component=self.unit, substrate=self.substrate
        )
. . .
```